### PR TITLE
Group DRO support for spectral risk uncertainty sets.

### PR DESCRIPTION
**Pull request by ronakdm**

### CODE REVIEW

File: deshift/_src/spectral_risk.py [hunk 1]
```diff
@@ -18,7 +18,8 @@ def make_spectral_risk_measure(
         which should be the same length as the batch size.
       penalty: either 'chi2' or 'kl' indicating which f-divergence 
         to use as the dual regularizer.
-      shift_cost: ehe non-negative dual regularization parameter.
+      shift_cost: the non-negative dual regularization parameter.
+      group_dist
```
Comment (vroulet) [inline deshift/_src/spectral_risk.py:22]: Remove group_dist

File: deshift/_src/spectral_risk.py [hunk 2]
```diff
@@ -33,6 +34,44 @@ def max_oracle(losses):
...
+ 
+    Args:
+      spectrum: a Numpy array containing the spectrum weights, 
+        which should be the same length as the number of groups.
+      penalty: either 'chi2' or 'kl' indicating which f-divergence 
+        to use as the dual regularizer.
+
+    Returns:
+      compute_sample_weight
+        a function that maps ``n`` losses to a vector of ``n`` weights on each training example.
```
Comment (vroulet) [inline deshift/_src/spectral_risk.py:52]: Add Reference paper as follows: 
```
References:
  Author, `Title <url>`_, year
```

File: deshift/_src/spectral_risk.py [hunk 3]
```diff
@@ -33,6 +34,44 @@ def max_oracle(losses):
...
+        assert torch.is_tensor(losses), "`losses` must be a PyTorch tensor"
+        assert torch.is_tensor(group_labels), "`group_labels` must be a PyTorch tensor"
+        assert torch.isin(torch.arange(group_labels.max() + 1), group_labels).all(), "all group labels must be observed"
+        with torch.no_grad():
+            device = losses.get_device()
+            device = device if device >= 0 else "cpu"
+
+            # count average loss of each group
+            unique_labels, labels_count = group_labels.unique(dim=0, return_counts=True)
+            res = torch.zeros_like(unique_labels, dtype=torch.float).scatter_add_(0, group_labels, losses)
```
Comment (vroulet) [inline deshift/_src/spectral_risk.py:64]: nitpick: as you named everything in a very readable manner (which makes reading and understanding the code a pleasure) maybe name "res" in a more explicit manner

File: tests/group_spectral_risk_test.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import sys
+sys.path.append(".")
+import deshift._src.spectral_risk as spectral_risk
+import deshift._src.spectra as spectra
+
+
+class GroupSpectralRiskTest(absltest.TestCase):
+
+  def test_spectral_risk(self):
+    test_func = spectral_risk.make_group_spectral_risk_measure
```
Comment (vroulet) [inline tests/group_spectral_risk_test.py:17]: Why do you use this test_func alias? Maybe avoid

File: tests/group_spectral_risk_test.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+import sys
+sys.path.append(".")
+import deshift._src.spectral_risk as spectral_risk
+import deshift._src.spectra as spectra
+
+
+class GroupSpectralRiskTest(absltest.TestCase):
+
+  def test_spectral_risk(self):
```
Comment (vroulet) [inline tests/group_spectral_risk_test.py:16]: Rather than making a loop below over "penalty", add penalty as an input of your test and use `absl.parameterized.product` (see below)
`absl.parameterized.product(a=[...], b=[...], ...)` runs the test for all given combinations of the inputs a, b, ... of your test. The advantage of using parameterized is that if the test fails it will tell you for which input it crashed. If you use a for loop you would not know (unless you add additional specific meessages etc...)
 
```
from absl.testing import parameterized
...
  @parameterized.product(penalty=["chi2", "kl"])
  def test_spectral_risk(self, penalty):
  ...
  # Remove then the for loop
```

**Review by vroulet — COMMENTED** [review]
Add the new function to the doc (see previous PR)
